### PR TITLE
[netcore] Fix Array.Copy* tests. Bump NETCORETESTS_VERSION version

### PR DESCRIPTION
**Pull request by EgorBo**
A small refactoring for Array.Copy/CopyConstrained. + reflect https://github.com/dotnet/coreclr/pull/25209 changes - new CoreFX tests (I updated `NETCORETESTS_VERSION `) have new Array tests with that new logic.

I tried to copy code from CoreCLR (see comments).

So now it's possible to copy int[] to uint[] and vise versa, IntPtr[] to UIntPtr[] (see comments in https://github.com/dotnet/coreclr/pull/25209).
Also, Array.cs doesn't use System.Reflection.DefaultBinder.cs any more.

### CODE REVIEW

File: mono/metadata/icall.c [hunk 1]
```diff
@@ -269,7 +269,7 @@ ves_icall_System_Array_GetValue (MonoArrayHandle arr, MonoArrayHandle indices, M
 void
 ves_icall_System_Array_SetValueImpl (MonoArrayHandle arr, MonoObjectHandle value, guint32 pos, MonoError *error)
 {
-	array_set_value_impl (arr, value, pos, FALSE, error);
+	array_set_value_impl (arr, value, pos, FALSE, TRUE, error);
```
Comment (EgorBo) [inline mono/metadata/icall.c]: While we share `array_set_value_impl ` between Array.SetValue and Array.Copy/CopyConstrained the later is more relaxed around primitive type conversions

File: netcore/Makefile [hunk 2]
```diff
@@ -160,7 +160,7 @@ run-tests-corefx-%: prepare update-tests-corefx
 	cp corefx/restore/corefx-restore.runtimeconfig.dev.json corefx/tests/extracted/$*/xunit.console.runtimeconfig.dev.json
 	cp corefx/restore/corefx-restore.runtimeconfig.dev.json corefx/tests/extracted/$*/Microsoft.DotNet.RemoteExecutorHost.runtimeconfig.dev.json
 	cp corefx/restore/corefx-restore.dll corefx/tests/extracted/$*/corefx-restore.dll
-	sed -i -e 's/9.9.9/$(NETCOREAPP_VERSION)/g' corefx/tests/extracted/$*/*.runtimeconfig.json
+	sed -i -e 's/3.0.0\"/$(NETCOREAPP_VERSION)\"/g' corefx/tests/extracted/$*/*.runtimeconfig.json
```
Comment (EgorBo) [inline netcore/Makefile]: NOTE: they use 3.0.0 as a placeholder in json files now (it was 9.9.9)

File: mono/metadata/icall.c [hunk 3]
```diff
@@ -279,8 +279,100 @@ set_invalid_cast (MonoError *error, MonoClass *src_class, MonoClass *dst_class)
...
+// Copied from CoreCLR: https://github.com/dotnet/coreclr/blob/030a3ea9b8dbeae89c90d34441d4d9a1cf4a7de6/src/vm/invokeutil.h#L176
+inline static 
+gboolean can_primitive_widen (MonoTypeEnum src_type, MonoTypeEnum dest_type)
+{
+	if (dest_type >= PRIMITIVE_TABLE_SIZE || src_type >= PRIMITIVE_TABLE_SIZE) {
+		if ((MONO_TYPE_I == dest_type && MONO_TYPE_I == src_type) ||
+			(MONO_TYPE_U == dest_type && MONO_TYPE_U == src_type)) {
+			return TRUE;
+		}
+		return FALSE;
```
Comment (jaykrell) [inline mono/metadata/icall.c]: Replace second `if` with `return`? I.e. return the expression?

File: mono/metadata/icall.c [hunk 4]
```diff
@@ -279,8 +279,100 @@ set_invalid_cast (MonoError *error, MonoClass *src_class, MonoClass *dst_class)
...
 
+#if ENABLE_NETCORE
+void
+ves_icall_System_Array_SetValueRelaxedImpl (MonoArrayHandle arr, MonoObjectHandle value, guint32 pos, MonoError *error)
+{
+	array_set_value_impl (arr, value, pos, FALSE, FALSE, error);
+}
+
+// Copied from CoreCLR: https://github.com/dotnet/coreclr/blob/d3e39bc2f81e3dbf9e4b96347f62b49d8700336c/src/vm/invokeutil.cpp#L33
+#define PRIMITIVE_TABLE_SIZE  MONO_TYPE_STRING
```
Comment (jaykrell) [inline mono/metadata/icall.c]: This `#define` isn't really needed, you can just say `static const guint32 primitive_conversions [ ] = { … }` without an explicit size, unless you know this is the size and are double checking yourself, that you didn't put in extra entries (too few entries won't be caught, just zero-padded).

File: mono/metadata/icall.c [hunk 5]
```diff
@@ -279,8 +279,100 @@ set_invalid_cast (MonoError *error, MonoClass *src_class, MonoClass *dst_class)
...
+	PT_Primitive | 0x3500,	// MONO_TYPE_I4   (W = I4, I8, R4, R8)
+	PT_Primitive | 0x3E00,	// MONO_TYPE_U4   (W = U4, I8, R4, R8)
+	PT_Primitive | 0x3400,	// MONO_TYPE_I8   (W = I8, R4, R8)
+	PT_Primitive | 0x3800,	// MONO_TYPE_U8   (W = U8, R4, R8)
+	PT_Primitive | 0x3000,	// MONO_TYPE_R4   (W = R4, R8)
+	PT_Primitive | 0x2000,	// MONO_TYPE_R8   (W = R8) 
+};
+
+// Copied from CoreCLR: https://github.com/dotnet/coreclr/blob/030a3ea9b8dbeae89c90d34441d4d9a1cf4a7de6/src/vm/invokeutil.h#L176
+inline static
```
Comment (jaykrell) [inline mono/metadata/icall.c]: Imho remove `inline`, let compiler decide.

File: mono/metadata/icall.c [hunk 6]
```diff
@@ -279,8 +279,100 @@ set_invalid_cast (MonoError *error, MonoClass *src_class, MonoClass *dst_class)
...
+	case MONO_TYPE_U2:
+	case MONO_TYPE_U4:
+	case MONO_TYPE_U8:
+	case MONO_TYPE_U:
+		return (MonoTypeEnum) (elementType - 1); // normalize to signed type
+	default:
+		break;
+	}
+
+	return elementType;
```
Comment (jaykrell) [inline mono/metadata/icall.c]: Remove `default`, or replace its `break` with `return`?

File: mono/metadata/icall.c [hunk 7]
```diff
@@ -446,6 +538,17 @@ array_set_value_impl (MonoArrayHandle arr_handle, MonoObjectHandle value_handle,
 	if (vt_isenum)
 		vt = mono_class_enum_basetype_internal (m_class_get_byval_arg (vc)->data.klass)->type;
 
+#if ENABLE_NETCORE
+	// Treat MONO_TYPE_U/I as MONO_TYPE_U8/I8/U4/I4
+#if SIZEOF_VOID_P == 8
```
Comment (jaykrell) [inline mono/metadata/icall.c]: This seems like a recurring pattern in mono to capture somehow, the mapping of `I4` or `I8` to native `I`, and `U`.

File: mono/metadata/icall.c (hunk 4 above)
Comment (EgorBo) [reply mono/metadata/icall.c]: @jaykrell `PRIMITIVE_TABLE_SIZE` is used inside `can_primitive_widen` to catch types that can not be widen or require special case ((U)IntPtr)

File: mono/metadata/icall.c (hunk 7 above)
Comment (EgorBo) [reply mono/metadata/icall.c]: I tried to find any similar patterns but with no luck

File: netcore/CoreFX.issues.rsp [hunk 8]
```diff
@@ -836,6 +845,9 @@
 ##  System.Reflection.Emit.Lightweight.Tests
 ####################################################################
 
+-nomethod System.Reflection.Emit.Tests.DynamicMethodctor1.String_Type_TypeArray_Type
+-nomethod System.Reflection.Emit.Tests.DynamicMethodctor1.String_Type_TypeArray_Module
+
 -nomethod System.Reflection.Emit.Tests.DynamicMethodGetILGenerator1.GetILGenerator_Int_Module_CoreclrIgnoresSkipVisibility
 -nomethod System.Reflection.Emit.Tests.DynamicILInfoTests.Test_TwoDimTest
 -nomethod System.Reflection.Emit.Tests.DynamicMethodToString.ToStringTest
```
Comment (marek-safar) [inline netcore/CoreFX.issues.rsp]: Why the fixed tests are not removed from the exclusion list?

Comment (EgorBo) [reply netcore/CoreFX.issues.rsp]: Those were unrelated failing SRE tests, I fixed them by backporting this CoreCLR PR https://github.com/dotnet/coreclr/pull/24937

**Comment by EgorBo** [comment]
@monojenkins build mono-coreonly-CI

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@marek-safar @steveisok can we merge it? It has 0 failures for netcore mode.
the failing lanes it seems are unrelated and presented in other PRs too, e.g. https://github.com/mono/mono/pull/15484

**Pull request merged**